### PR TITLE
Improved search space XGBoost [Continue]

### PR DESCRIPTION
**Pull request by jeromepatel**
## Motivation
This PR is with reference to issue #1683, and the continuation of my previous work of PR #2322.

## Description of the changes

Unfortunately, I made changes to <code>master</code> branch which was not the right thing to do, Sorry for that. Apart from that I have described all the changes in PR #2322, which can be summarized as, added new parameters
 <code> subsample,colsample_bytree,tree_method,min_child_weight</code> tuned them according to <code> accuracy </code> and several comments were added to the example.

### CODE REVIEW

File: examples/xgboost_cv.py [hunk 1]
```diff
@@ -1,3 +1,4 @@
+<<<<<<< Updated upstream
```
Comment (nzw0301) [inline examples/xgboost_cv.py:1]: Could you remove this line? I think that it comes from the git-conflict.

File: examples/xgboost_cv.py [hunk 2]
```diff
@@ -91,3 +92,104 @@ def objective(trial):
     print("  Number of estimators: {}".format(trial.user_attrs["n_estimators"]))
 
     shutil.rmtree(CV_RESULT_DIR)
+=======
```
Comment (nzw0301) [inline examples/xgboost_cv.py:95]: ditto

File: examples/xgboost_cv.py [hunk 3]
```diff
@@ -91,3 +92,104 @@ def objective(trial):
...
+
+    print("  Value: {}".format(trial.value))
+    print("  Params: ")
+    for key, value in trial.params.items():
+        print("    {}: {}".format(key, value))
+
+    print("  Number of estimators: {}".format(trial.user_attrs["n_estimators"]))
+
+    shutil.rmtree(CV_RESULT_DIR)
+>>>>>>> Stashed changes
```
Comment (nzw0301) [inline examples/xgboost_cv.py:195]: ditto

File: examples/xgboost_simple.py [hunk 4]
```diff
@@ -1,3 +1,4 @@
+<<<<<<< Updated upstream
```
Comment (nzw0301) [inline examples/xgboost_simple.py:1]: ditto

File: examples/xgboost_simple.py [hunk 5]
```diff
@@ -63,3 +64,84 @@ def objective(trial):
     print("  Params: ")
     for key, value in trial.params.items():
         print("    {}: {}".format(key, value))
+=======
```
Comment (nzw0301) [inline examples/xgboost_simple.py:67]: ditto

File: examples/xgboost_simple.py [hunk 6]
```diff
@@ -63,3 +64,84 @@ def objective(trial):
...
+
+    print("Number of finished trials: ", len(study.trials))
+    print("Best trial:")
+    trial = study.best_trial
+
+    print("  Value: {}".format(trial.value))
+    print("  Params: ")
+    for key, value in trial.params.items():
+        print("    {}: {}".format(key, value))
+>>>>>>> Stashed changes
```
Comment (nzw0301) [inline examples/xgboost_simple.py:147]: ditto

**Comment by jeromepatel** [comment]
I have successfully migrated changes from last PR Pull #2322, formated changes for tests in this PR. If there are any changes need to be made, please let me know. Thank you.

**Comment by HideakiImamura** [comment]
@toshihikoyanase Let me assign you since you reviewed #2322. Please feel free to remove the assignment if you are busy.

File: examples/xgboost_cv.py [hunk 7]
```diff
@@ -1,93 +1,95 @@
...
-    print("  Value: {}".format(trial.value))
-    print("  Params: ")
-    for key, value in trial.params.items():
-        print("    {}: {}".format(key, value))
-
-    print("  Number of estimators: {}".format(trial.user_attrs["n_estimators"]))
-
-    shutil.rmtree(CV_RESULT_DIR)
+"""
+Optuna example that optimizes a classifier configuration for cancer dataset using XGBoost.
```
Comment (nzw0301) [inline examples/xgboost_cv.py:2]: Could you add a blank line after this sentence like the previous version?

Comment (jeromepatel) [reply examples/xgboost_cv.py:2]: okay, sure

**Comment by Crissman** [comment]
Please resolve the conflicts. 🙇

**Comment by jeromepatel** [comment]
@Crissman  Thanks for reviewing. I have resolved conflicts when merging with `optuna/master`. PTAL

File: examples/xgboost/xgboost_simple.py [hunk 8]
```diff
@@ -1,65 +1,79 @@
...
+import sklearn.datasets
+import sklearn.metrics
+from sklearn.model_selection import train_test_split
+import xgboost as xgb
+
+import optuna
+
+
+# FYI: Objective functions can take additional arguments
+# (https://optuna.readthedocs.io/en/stable/faq.html#objective-func-additional-args).
```
Comment (Crissman) [inline examples/xgboost/xgboost_simple.py:21]:
```suggestion
```

If we don't use additional arguments here, I don't think it's useful to mention it's possible.

**Review by Crissman — CHANGES_REQUESTED** [review]
Some doc edits.
(Note to self: do test run and compare to other examples.)

File: examples/xgboost/xgboost_simple.py (hunk 8 above)
Comment (jeromepatel) [reply examples/xgboost/xgboost_simple.py:21]: Thanks for your comment. I agree with you that the lines seem redundant related to this example. It was in the older example before I modified it, but anyway I'll remove that.

**Review by Crissman — APPROVED** [review]
LGTM!

**Comment by Crissman** [comment]
@nzw0301 Can you approve the changes, please?

**Review by nzw0301 — APPROVED** [review]

**Comment by Crissman** [comment]
Tested CI tests for examples for this PR. No issue. 👍

**Comment by Crissman** [comment]
@jeromepatel It looks like the entire file was deleted and re-pasted, so even lines that are not changed are showing up as changed.

Can you rework this PR so that only the changes are highlighted?

**Comment by jeromepatel** [comment]
@Crissman I think it was due to git conflict from my earlier commits or probably from merge conflict after the example folder changes, I am not sure. I am still looking for solutions to resolve this. If you could suggest a solution, it would be helpful. As a final solution, I can close this PR and submit a new one, with corresponding changes in the `xgboost_simple.py` file only which may resolve all the issues related to merge/conflict.

**Comment by nzw0301** [comment]
Like this?
```bash
git clone git@github.com:optuna/optuna.git
git remote rename origin upstream
git checkout -b xgboost-example
git remote add origin git@github.com:jeromepatel/optuna.git
# update the changed files in this PR
git add .
git commit
git push -f origin xgboost-example
```

**Comment by jeromepatel** [comment]
Thank you @nzw0301 for helping throughout the PR, my new commit reflects a new checkout branch changes with a clear difference. Thank you for reviewing @Crissman

**Comment by nzw0301** [comment]
Great! Will `xgboost_cv.py` be updated later?

**Comment by jeromepatel** [comment]
I am looking forward to updating `xgboost_cv.py` too. I have some suggestions about optimizing GANs using Optuna (in Keras and Pytorch), as GANs are becoming popular and they are difficult to tune as well.

**Comment by Crissman** [comment]
Thanks for the git cleanup, Jerome. Please add the changes you had previously to `xgboost_cv.py` to this PR as well. 🙇

**Comment by jeromepatel** [comment]
@Crissman Thank you for your review. As I was adding changes to `xgboost_cv.py` I also have modified the hyperparameter settings from my previous PR #2322 which I didn't previously plan to append in this PR. PTAL.

**Comment by Crissman** [comment]
Approved. @toshihikoyanase PTAL

**Comment by toshihikoyanase** [comment]
@Crissman Thank you for letting me know. As you and @nzw0301 have already approved this PR, I think you can merge it.

Regarding the issue #1683, I think we may make the experimental settings closer to the real-world applications if we change the dataset. So, how about keeping it open?

@nzw0301 Thank you for your review. Let me fix the assignees.

**Comment by nzw0301** [comment]
@jeromepatel Thank you for improving the examples related to XGBoost. I think that the examples give more informative explanations of the hyper-parameters now. Let me merge this PR!

I agree with @toshihikoyanase, I have locally compared validation accuracy between this PR's `xgboost_simple.py` and the master branch's `xgboost_simple.py`,  and their validation accuracies are almost the same. I think that it might be difficult to improve performance anymore for this example problem because the objective’s performance on the master branch is supposed to be saturated (validation accuracy was 98%).

---

I ran `xgboost_simple.py` 10 times for each master and PR brach.

```
master branch -- mean: 0.9832167832167832
           PR -- mean: 0.9825174825174827
```

The results on the `digit` dataset that is a 10 class classification problem:

```
master branch -- mean: 0.966  std:0.009111111111111117
           PR -- mean: 0.9668 std:0.008459051693633013
```

**Comment by nzw0301** [comment]
Let me change the title to remove `[Continue]`.
